src/log.py

```diff
@@ -0,0 +1,55 @@
+"""
+Data Log
+
+By: Norma Chang
+"""
+import json
+import time
+
+def log(data, data_type):
+    """
+    Logs dictionary data from leader
+    """
+    if data_type == 'leader':
+        leader_dict = {
+            'Time' : time.time(), 
+            'Input' : data._distance,
+            'Speed' : data._speed,
+        }
+        write_to_file(leader_dict, data_type)
+    elif data_type == 'follower':
+        follower_dict = {
+            'Time' : time.time(), 
+            'Distance' : data._distance,
+            'Speed' : data._speed,
+            'Turn Angle' : data._turn_angle,
+            'Yaw' : data._yaw,
+        }
+        write_to_file(follower_dict, data_type)
+    else:
+        pass
+
+def write_to_file(data, file_name):
+    """
+    Write directionary data to text file.
+    """
+    with open(file_name + '.log', 'a') as outfile:
+        json.dump(data, outfile)
+        outfile.write('\n')
+
+def array_dict(file_name):
+  """
+  Returns the log file containing the cycle data.
+  """
+  string_arr = []
+  array_dict = []
+
+  log_file = open(file_name, 'r')
+  log = log_file.read().strip().split('\n')
```

Comment:
This whole function can be one line of code maximum if we had commas in the log file.

Reply:
Same thing as splitting by newline

Reply:
It can definitely be shorter if you split it by comma. Exactly one line of code total to read the file actually, while being cleaner, faster, and more memory efficient at the same time.

Reply:
There's a \n at the end of every line already, it saves me the trouble of appending commas when writing, removing them when reading while still leaving the new lines vs. this way, it cleans up the new lines from the end of each string. It might be one line but its not more efficient or cleaner. Its less efficient because you have to now clean up the commas and the newlines when reading.

Reply:
That's not the point. The way you're parsing this is by reading the file (iterates through the whole file), stripping spaces (iterates through the whole file again) and splitting by \n (reads the whole file again). Then you iterate through each element and append it into another array (for some reason). So now we're at O(4n) time and O(2n) space. You're then iterating through the entire array you created and loading each element as a json object. So now we're at O(4n+(n*m)) run time and O(3n) space. It's definitely not cleaner nor efficient than appending the comma already as you're writing the dictionaries into the files. Because when it comes time to read it back, it'll already be in a format that json.loads(file) will pick up on the spot.

Reply:
So maybe this is where the misunderstanding is. For reading, stripping and splitting, its only opening the file once, it doesnt open read, close, open strip, close then open split, close.
It opens the file once, loads the entire file into a string, separates the long string into substrings by splitting by new line, converts each substring into a dictionary then appends the dictionaries into an array as requested.

Reply:
So my process so we're both on the same page:
It opens the file once, loads the entire file into a string, separates the long string into substrings by splitting by new line, converts each substring into a dictionary then appends the dictionaries into an array as requested.

- You cant parse multiple json objects into an array of dictionaries without the loop
- You cant just write the data into an array of dictionaries as you do not know where the end bracket would go in a continuous log (defining attribute is that doesnt have an end)
"Because when it comes time to read it back, it'll already be in a format that json.loads(file) will pick up on the spot." <-- is only possible if you know where the ending bracket goes in the first place

Reply:
You basically just explained why your algorithm is very inefficient. It’s iterating through the exact same data multiple times until it can eventually be able to return the information to us. There’s no need to do any of this. 

 As to address the rest of the concerns, you don’t even need to write the data into an array, then an array of dictionaries, and creating a JSON object of each dictionary. JSON already takes care of that with the loads function. 

And for the record, it is very possible to know where the brackets go in the log (hint, at the beginning of the string and at the end of the string). Now we can have an array without having to store anything. Pass the string you read into the loads function with opening and closing brackets on both ends of the string and it’ll pick it up and makes it a whole JSON object on the spot. One line of code. 

Here’s a link to run time complexities in Python for more information: https://wiki.python.org/moin/TimeComplexity

Reply:
The read function does not iterate through the file multiple times because log_file.read().strip().split('\n') is not in a loop and it reads the entire file in as one string once during the .read(). Afterwards strip removes whitespace and splits the string by newline in memory. 

As for the last part, dictionary = json.loads(string) takes the string from what I just read into loads. This entire read function doesn't run when you start the leader/follower, its literally just there to give you the [{'key': 'value'}, {'key': 'value'}, {'key': 'value'}] format that was requested. Only the write function runs and all it does is write the dictionary data to a file. 

Perhaps I didnt explain this well the first time, so while the log is running I have this: 
{'key': 'value'},
{'key': 'value'},
{'key': 'value'}

That is an invalid JSON array but valid JSON objects. Now I need to put brackets to make it a valid array so if I put a beginning bracket:
[{'key': 'value'},
{'key': 'value'},
{'key': 'value'}
{'key': 'value'}......

Where do I put the ending bracket if I dont know where the end is? Putting a bracket at the end of each line and deleting when new lines are written? That would be inefficient. 

Instead to solve that problem, I load each JSON object into a python array in order to generate an array of dictionaries in [{'key': 'value'}, {'key': 'value'}, {'key': 'value'}] format. 

From this point for interpretation use, all you need to do is call array_of_dict() function which will give you the array that you need. Please look at my latest commit, I updated it with more comments that should be more thorough than my initial commit.

Reply:
What do you think your code is doing under the hood? You clearly don't understand what that line is doing. Let me break it down for you.

read() -> Loads the whole text into a string. It does so by **iterating** over every character in the file and builds a string out of it. Time Complexity: O(n). Space Complexity: O(n).

strip() -> Removes the empty white space from the string. This is done by **iterating** through the string to find the white spaces. It's not done in place so it needs to be copied into a temp variable then back into log while log is being used to iterate over the entire file. Time Complexity: O(n). Space Complexity: O(n). 

split('\n') -> Splits the string into an array by new line character. This is done by **iterating** through the string to find the '\n'. Takes everything before that and appends it to an array. It creates a temporary array while log is being used to iterate over the entire file. Time Complexity: O(n). Space Complexity: O(n). 

Your line of code is the same thing as
log = log_file.read() 
log = log.strip() 
log = log.split('\n')

You're running through the same contents 3 times in this line of code alone. Absolutely nothing is done all at once. I'm not going to keep going for the rest of the file. Check the runtime I described in my comments above.

I know the read function doesn't run in the leader. You don't need to store the brackets in the file at all per my previous comment. You will load the brackets **on top** of string that you're reading from the file. Dig this:

log.json
{'key': 'value'}, {'key': 'value'}, {'key': 'value'}

Read it as a file
log = "{'key': 'value'}, {'key': 'value'}, {'key': 'value'}"

Pass it into JSON

log = json.loads('['+log+']')

now JSON will parse a string that looks like this: 
"[{'key': 'value'}, {'key': 'value'}, {'key': 'value'}]"

And this will make it a **valid** JSON format, and the whole file is parsed correctly for us to use. 

I just did this whole function in one line of code by iterating through only once for reading, once for converting into JSON. 

P.S. I asked you to change the name array_of_dict 8 commits ago because it's not descriptive at all.